aws/sdk-codegen/src/main/kotlin/software/amazon/smithy/rustsdk/AwsPresigningDecorator.kt

```diff
@@ -155,7 +154,7 @@ class AwsInputPresignedMethod(
         }
 
     private fun RustWriter.writeInputPresignedMethod(section: OperationSection.InputImpl) {
-        val operationError = operationShape.errorSymbol(symbolProvider)
+        val operationError = symbolProvider.symbolForOperationError(operationShape)
```

Comment:
The problem arises that Smithy doesn't really have a concept of an operation error. The operation has errors, but the fact that there is an operation error type is an implementation detail of smithy-rs and the SDK. So that `operation.error()` would need to return something non-Smithy if we were to take that approach. Other than that, things like `toSymbol()` will need to take a symbol provider as an argument, or alternatively we delve into some questionable global variable practices. There are certain Smithy patterns that will get more strict over time rather than less strict.

Something we can consider is making an abstraction layer on top of Smithy, but it would be expensive to refactor to that, and it would make Smithy upgrades more expensive in the future.

Reply:
I spoke with John about this outside of GitHub. Including some notes here for posterity:

## My assumptions

**Smithy** is a modelling language for defining a **service** that responds to requests.
- A **model** defines data structures representing requests, responses, and errors.
- An **operation** is a data structure that associates a request, response, and a zero or more errors.
- Requests, responses, and errors are defined as objects containing key-value pairs. Values may be aggregate types (like maps and lists) or primitive types. Collectively, objects and types are referred to as "shapes."
- Shapes may be annotated with one or more **traits** that influence codegen.
- During codegen, we may create **synthetic** shapes and traits and attach them to the model. We must do this in cases where expected behavior isn't modeled. Additionally, we may do this in cases where we need to share information between individual codegen components or where modeled data must be transformed or augmented before it can be converted into Rust code.
- To generate code, shapes must be converted into **symbols** by a **symbol provider**. The symbol provider ensures no naming or keyword collisions occur when generating code.
- We process smithy models to generate a client. The client allows users to construct **operation input structs**, which are then converted into an **HTTP request** and sent. When an **HTTP response** is received, it is converted into either an **error struct** or an **operation output struct**.
- Structs are a Rust concept, not a Smithy one. Because Rust does not support union types, and because an **operation** in Smithy may be associated with 0 or more possible errors, we aggregate those errors into an opaque **error struct** which must be assigned a name by the **symbol provider**.

## Notes from our discussion

- `RuntimeType` shouldn't really exist. We should instead merge the good parts of it into the `Symbol` class.
- `smithy-rs` is not a general-purpose library for writing Kotlin programs that generate Rust. Rather, it's a tool for converting smithy models in Rust code that's usable by our runtime crates.
- Symbols represent types, but we shouldn't extend them to represent syntactic things like _"a block of code"_ or _"a variable name."_ Instead, we should decide whether we need an abstraction to represent those things or should stick to formatting strings.
- We should devote a small amount of time to investigating cases where pushing back the "writing a symbol into a string" phase of codegen would make codegen simpler to extend and maintain. We already do this with the "section" customization APIs. Do we need anything more?